contracts/lib/OnchainNonMerklizedIdentityBase.sol

```diff
@@ -0,0 +1,245 @@
+import {IdentityLib} from "../lib/IdentityLib.sol";
+import {IdentityBase} from "../lib/IdentityBase.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+
+/**
+ * @dev Contract building onchain non-merklized identity
+ * that can be transformed to W3C verifiable credential
+ */
+abstract contract OnchainNonMerklizedIdentityBase is IdentityBase {
+    using IdentityLib for IdentityLib.Data;
+
+    /**
+     * @dev CredentialInformation keep information about the credential
+     */
+    struct CredentialInformation {
+        string[] jsonLDContextUrls;
+        string jsonSchemaUrl;
+        string _type;
+    }
+
+    /**
+     * @dev CredentialMetadata keep metadata about the credential
+     */
+    struct CredentialMetadata {
+        uint256 id;
+        uint64 revocationNonce;
+        uint64 issuanceDate;
+        uint64 expirationDate;
+    }
+
+    /**
+     * @dev Claim keep information about claim and calculated hashIndex and hashValue
+     */
+    struct Claim {
+        uint256[8] coreClaim;
+        uint256 hashIndex;
+        uint256 hashValue;
+    }
+
+    /**
+     * @dev SubjectField credential subject for the credential
+     * key - name of the field
+     * value - value of the field
+     * rawValue - raw value of the field,
+     * is used for string and double types to restore source value in W3C verifiable credential
+     */
+    struct SubjectField {
+        string key;
+        uint256 value;
+        bytes rawValue;
+    }
+
+    /**
+     * @dev Status information about the credential status
+     */
+    struct Status {
+        string id;
+        string _type;
+        uint64 revocationNonce;
+    }
+
+    /**
+     * @dev Schema information about the credential schema
+     */
+    struct Schema {
+        string id;
+        string _type;
+    }
+
+    /**
+     * @dev IssuerState issuer state and roots
+     */
+    struct IssuerState {
+        uint256 rootOfRoots;
+        uint256 claimsTreeRoot;
+        uint256 revocationTreeRoot;
+        uint256 value;
+    }
+
+    /**
+     * @dev IssuerData issuer data
+     */
+    struct IssuerData {
+        uint256 id;
+        IssuerState state;
+    }
+
+    /**
+     * @dev IssuanceProof information about the issuance proof
+     */
+    struct IssuanceProof {
+        string _type;
+        uint256[8] coreClaim;
+        IssuerData issuerData;
+        SmtLib.Proof mtp;
+    }
+
+    /**
+     * @dev CredentialData information about the credential.
+     * Can be transformed to W3C verifiable credential
+     */
+    struct CredentialData {
+        string id;
+        string[] context;
+        string[] _type;
+        uint64 expirationDate;
+        uint64 issuanceDate;
+        uint256 issuer;
+        SubjectField[] credentialSubject;
+        Status credentialStatus;
+        Schema credentialSchema;
+        IssuanceProof[] proof;
+    }
+
+    /**
+     * @dev Id
+     */
+    struct Id {
```

Comment:
Maybe it's better to remove this struct and just use uint256.

Reply:
Fixed